packages/next/src/server/app-render/create-component-tree.tsx

```diff
@@ -564,71 +572,88 @@ async function createComponentTreeInternal({
+      // a lazy hole rather than null
+      <Segment
+        key={cacheNodeKey}
+        isStaticGeneration={isStaticGeneration}
+        ready={getMetadataReady}
+      >
         {layerAssets}
         <Component {...props} />
-      </>
-    )
+      </Segment>,
+      parallelRouteCacheNodeSeedData,
+      loadingData,
+    ]
   }
-
-  return [
-    actualSegment,
-    <React.Fragment key={cacheNodeKey}>
-      {segmentElement}
-      {/* This null is currently critical. The wrapped Component can render null and if there was not fragment
-            surrounding it this would look like a pending tree data state on the client which will cause an error
-            and break the app. Long-term we need to move away from using null as a partial tree identifier since it
-            is a valid return type for the components we wrap. Once we make this change we can safely remove the
-            fragment. The reason the extra null here is required is that fragments which only have 1 child are elided.
-            If the Component above renders null the actual tree data will look like `[null, null]`. If we remove the extra
-            null it will look like `null` (the array is elided) and this is what confuses the client router.
-            TODO-APP update router to use a Symbol for partial tree detection */}
-      {null}
-    </React.Fragment>,
-    parallelRouteCacheNodeSeedData,
-    loadingData,
-  ]
 }
 
 async function MetadataOutlet({
-  getReady,
+  ready,
 }: {
-  getReady: () => Promise<void> & { status?: string; value?: unknown }
+  ready: () => Promise<void> & { status?: string; value?: unknown }
 }) {
-  const ready = getReady()
-  // We actually expect this to be an instrumented promise and once this file is properly
-  // moved to the RSC module graph we can switch to using React.use for this synchronous unwrapping.
-  // The synchronous unwrapping will become important with dynamic IO since we want to resolve metadata
-  // before anything dynamic can be triggered
-  if (ready.status === 'rejected') {
-    throw ready.value
-  } else if (ready.status !== 'fulfilled') {
-    await ready
+  const r = ready()
+  // We can avoid a extra microtask by unwrapping the instrumented promise directly if available.
+  if (r.status === 'rejected') {
+    throw r.value
+  } else if (r.status !== 'fulfilled') {
+    await r
   }
   return null
 }
+
+async function Segment({
+  isStaticGeneration,
+  ready,
+  children,
+}: {
+  isStaticGeneration: boolean
+  ready?: () => Promise<void>
+  children: React.ReactNode
+}) {
+  if (isStaticGeneration && ready) {
+    // During static generation we wait for metadata to complete before rendering segments.
+    // This is slower but it allows us to ensure that metadata is finished before we start
+    // rendering the segment which can synchronously abort the render is certain circumstances
+    try {
+      await ready()
+    } catch (e) {
```

Comment:
Errors not used can be excluded:

```suggestion
    } catch {
```